README.md
![pipeline](https://gitlab.com/incubateur-pe/kubernetes-bare-metal/badges/dev/pipeline.svg)

kubernetes-bare-metal
=========

Install a kubernetes cluster with kubeadm
The goal of this role is to provide a way to install a full featured, production ready cluster, in a single role.
This role is usable behind a corporate proxy, and/or through mirrors for accessing packages and images resources, you can use it in a totally cloisoned environment, without any direct access to the internet.

Prerequisites
------------

__Local__:

Python : see requirements.yaml : `pip install -r requirements.yaml`

__Servers__:

  * CRI : the runtime will be detected by kubeadm
  * Swap : No swap should be activated on the servers

__Inventory__:

  * Group kubernetes_masters : Contains all hosts which role will be kubernetes master
  * Group kubernetes_workers : Contains all hosts which role will be kubernetes workers


Variables
--------------

__repositories :__

| variable | scope | description | defaut |
| --- | --- | --- | --- |
| kubernetes_apt_repository | role | Kubernetes apt repository address | https://apt.kubernetes.io |
| kubernetes_apt_key_url | role | Url serving the google cloud public signing key | https://packages.cloud.google.com/apt/doc/apt-key.gpg |
| kubernetes_yum_repo | role | Kubernetes rpm repository address for RHEL | https://packages.cloud.google.com/yum/repos/ |
| kubernetes_yum_enable_gpg | role | Set to false to disable gpg check on yum repo | true |
| kubernetes_yum_repo_key | role | yum repository gpg key (repository) | https://packages.cloud.google.com/yum/doc/yum-key.gpg |
| kubernetes_yum_package_key | role | yum repository gpg key (packages) | https://packages.cloud.google.com/yum/doc/rpm-package-key.gpg |

__kubernetes cluster:__:

| variable | scope | description | defaut |
| --- | --- | --- | --- |
| kubernetes_version | role | Kubernetes version to install or upgrade to | 1.20.0 |
| node_name | host | Kubernetes node name, must be unique | see kubeadm documentation |
| kubernetes_cluster_name | role | Kubernetes cluster name | kubernetes |
| kubernetes_control_plane_endpoint | role | Load balancer (or apiServer if only one master) address serving the master nodes (dns name or ip) | first master's ansible_fqdn |
| kubernetes_control_plane_port | role | Load balancer port (or apiServer if only one master) of the apiServer | 6443 |
| kubernetes_image_repository | role | kubernetes images repository | k8s.gcr.io |
| kubernetes_api_server_advertise_address | host | Advertise address of the api server (example: 172.16.10.10 ) | ansible_default_ipv4.address |
| kubernetes_api_server_port | role | Api server listen port | 6443 |
| kubernetes_kubeconfig_file | role | Kubernetes local configuration file (fetched from the master) | {{ role_path }}/files/admin.conf |
| kubernetes_kube_proxy_metrics_bind_address | role | kube-proxy bind address, necessary to expose metrics for example | see kubeadm documentation |
| kubernetes_kube_controller_manager_bind_address | role | kube-controller-manager bind address | see kubeadm documentation |
| kubernetes_kube_controller_manager_port | role | kube-controller-manager insecure port (deprecated), necessary to expose metrics for example | N/A |
| kubernetes_kube_scheduler_bind_address | role | kube-scheduler bind address | see kubeadm documentation |
| kubernetes_kube_scheduler_port | role | kube-scheduler insecure port (deprecated), to expose metrics for example | N/A |
| kubernetes_cgroup_driver | role | cgroup driver to use for kubelet, default to auto detection by kubeadm | N/A |
| kubernetes_kubeadm_target | role | Playbook host that will be used as a target for all kubeadm/kubectl operations | first master |


__kubernetes networking:__:

| variable | scope | description | defaut |
| --- | --- | --- | --- |
| kubernetes_cni_driver | role | Defines the cni driver to use (currently supported : flannel, calico, antrea) | flannel |
| flannel_image | role | Name of the flannel repository/image | quay.io/coreos/flannel |
| flannel_version | role | Flannel image version to use | v0.13.0 |
| flannel_backend_type | role | Flannel backend to use | vxlan |
| calico_ip_autodetect_method | role | IP_AUTODETECTION_METHOD method value | first-found |
| calico_network_config | role | calico network configuration | see default/main.yml |
| kubernetes_proxy_mode | role | Kube-proxy mode (example: iptables, ipvs...) | ipvs |
| kubernetes_proxy_mode_ipvs_sync_period | role | Max duration between 2 ipvs syncs | 30s |
| kubernetes_proxy_mode_ipvs_min_sync_period | role | Min duration betwwen 2 ipvs syncs | 2s |
| kubernetes_proxy_mode_ipvs_scheduler | role | Ipvs scheduler to use ( rr,lc,dh,sh,sed,nq ) | rr |
| kubernetes_proxy_mode_iptables_sync_period | role | Max duration between 2 iptables syncs | 30s |
| kubernetes_proxy_mode_iptables_min_sync_period | role | Min duration betwwen 2 iptables syncs | 2s |
| kubernetes_pod_subnet | role | Subnet to configure as the podCIDR for the cni | 10.244.0.0/16 |
| kubernetes_service_subnet | role | Services subnet | voir documentation kubeadm |

__certificates :__

| variable | scope | description | defaut |
| --- | --- | --- | --- |
| kubernetes_pki.certificate | role | Pem file to upload and use for the cluster certificate authority | N/A |
| kubernetes_pki.private_key | role | Key file to upload and use for the cluster certificate authority | N/A |
| kubernetes_pki.ca_chain | role | Full CA chain for kube-controller-manager and pods | N/A |
| kubernetes_client_ca_file | role | Pem file to upload and use for the cluster client certificate chain | N/A |

__miscellaneous :__

| variable | scope | description | defaut |
| --- | --- | --- | --- |
| metrics_server_image | role | Repository/image to use for metrics_server | k8s.gcr.io/metrics-server/metrics-server |
| metrics_server_image_version | role | Version of metrics server image to use | v0.4.0 |
| metrics_server_wait_deploy | role | Wait the deployment of metrics-server, useful if you deploy the masters in a separate play | true |
| kubernetes_cloud_provider | role | Cloud provider to use | N/A |
| kubernetes_upgrade_consistency_check | role | Check for cluster nodes versions consistency ( set to false if upgrade is done with 2 plays) | true |
| kubernetes_upgrades_allowed | role | Should the role handle upgrades | true |
| kubernetes_confirm_upgrade | role | If an upgrade is needed, ask for user input before execution | false |

Certificates Management
-----------------------

The variables in the `certificates` section allows configuration of a custom CA for the cluster, for example, an intermediate CA, generated from a corporate internal PKI, which in turn will enable all client workstations to trust the cluster. This impacts the kubeconfig file for example, removing the need for the certificate-authority-data configuration

if you want to enable this behaviour, set `kubernetes_pki.certificate` and `kubernetes_pki.private_key` to the AC files, they will be uploaded to the masters servers and used as the cluster's CA

When a custom CA is provided, non self signed certificates are generated for kube-controller-manager and kube-scheduler

__intermediate ca__:

If the files provided aren't a root CA, you should provide the full ca chain file and set `kubernetes_pki.ca_chain` accordingly.
Also pay attention to the pathlen size of your CA if you plan to generate etcd ca from this CA

Playbook Example
----------------

```yaml
- hosts: all
  remote_user: root
  roles:
    - kubernetes-bare-metal
  vars:
    kubernetes_control_plane_endpoint: master.k8s.loc
    kubernetes_cluster_name: dev
    k8s_version: 1.20.0
```

It remains possible to dissociate masters and workers installation, although the masters must be in the same ansible inventory:

```yaml
- hosts: kubernetes_masters
  remote_user: root
  roles:
    - kubernetes-bare-metal
  vars:
    kubernetes_control_plane_endpoint: master.k8s.loc
    kubernetes_cluster_name: dev
    k8s_version: 1.20.0
    metrics_server_wait_deploy: false # Metrics-server will be waited in the worker play

- hosts: kubernetes_workers
  remote_user: root
  roles:
    - kubernetes-bare-metal
  vars:
    kubernetes_control_plane_endpoint: master.k8s.loc
    kubernetes_cluster_name: dev
    k8s_version: 1.20.0
    kubernetes_upgrade_consistency_check: no # Déactivate versions consistency check as part of the cluster may already be up to date
```

Tests
-----

The tests use molecule + libvirt + kvm distant + testinfra, allowing on the fly multi-nodes clusters creation and system configuration checks

Install prerequisites : `pip install -r requirements-tests.yaml`
Install vagrant : https://www.vagrantup.com/downloads
Install vagrant-libvirt plugin : `vagrant plugin install vagrant-libvirt`

__molecule/libvirt issues:__

  * [issue #921](https://github.com/vagrant-libvirt/vagrant-libvirt/issues/921#issuecomment-464334757) : Requires to edit the plugin on certain os
  * The vagrant port-forward isn't used by molecule, acces is made directly to the VM management interface, so the management network must be routed
    * Workaround: use OpenSSH option ProxyCommand in molecule config

__Tests execution with testInfra:__

With a local KVM/Qemu, everything works fine, although distant vms access on a distant KVM host, if the network isn't routed requires an ssh config like :

```
Host bastion
  Hostname kvm
  IdentityFile ~/.ssh/id_rsa
  User libvirt-user

Host 192.168.121.*
  ProxyJump bastion
```

__Tests execution with docker:__
```
# interactive :
docker run --env-file molecule/commun/env_local_tests.sh -v $(pwd):/sources/kubernetes-bare-metal -w /sources/kubernetes-bare-metal   -v ~/.vagrant.d/boxes/:/root/.vagrant.d/boxes/   -v /var/run/libvirt/libvirt-sock:/var/run/libvirt/libvirt-sock --dns 192.168.121.1 --dns 10.0.4.1 --network host   -it --entrypoint bash ulrichg/molecule-vagrant-libvirt:latest

# lint:
docker run --env-file molecule/commun/env_local_tests.sh -v $(pwd):/sources/kubernetes-bare-metal -w /sources/kubernetes-bare-metal   -v ~/.vagrant.d/boxes/:/root/.vagrant.d/boxes/   -v /var/run/libvirt/libvirt-sock:/var/run/libvirt/libvirt-sock ulrichg/molecule-vagrant-libvirt:latest lint

# default scenario:
docker run --env-file molecule/commun/env_local_tests.sh -v $(pwd):/sources/kubernetes-bare-metal -w /sources/kubernetes-bare-metal   -v ~/.vagrant.d/boxes/:/root/.vagrant.d/boxes/   -v /var/run/libvirt/libvirt-sock:/var/run/libvirt/libvirt-sock --dns 192.168.121.1 --dns 10.0.4.1 --network host  ulrichg/molecule-vagrant-libvirt:latest

# other scenario:
docker run --env-file molecule/commun/env_local_tests.sh -v $(pwd):/sources/kubernetes-bare-metal -w /sources/kubernetes-bare-metal   -v ~/.vagrant.d/boxes/:/root/.vagrant.d/boxes/   -v /var/run/libvirt/libvirt-sock:/var/run/libvirt/libvirt-sock --dns 192.168.121.1 --dns 10.0.4.1 --network host  ulrichg/molecule-vagrant-libvirt:latest test -s cluster1master
```
License
-------

BSD 3-Clause


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
### Changes


## 1.0.8
### Changes
- Add antrea CNI deployment
- define new variable kubernetes_kubeadm_target

### Fixed
- kubernetes mandatory modules are now loaded on boot
- kube-controller-manager is fully working when custom ca_chain is defined

## 1.0.5
### Changes
- Set kubernetes default version to 1.21.0

### Fixed
- Apt key on debian os is now fetched from google cloud

## 1.0.4
### Changes
- Update calico to v3.18.1


## 1.0.3
### Changes
- Update calico to v3.17.3

### Fixed
- client_ca pem file access rights
- copy all custom ca certificates when multiple masters

## 1.0.2
### Added
 - ability to change ip autodetect method

## 1.0.1
### Added
- Ability to define a metrics bind address for kube-proxy
- Ability to define a bind address for kube-controller-manager
- Ability to define a bind address for kube-scheduler
- Ability to define a cgroup driver
- When custom AC is defined, use it to generate cert and key for kube-controller-manager
- When custom AC is defined, use it to generate cert and key for kube-scheduler
- Supports kubernetes version 1.20.0
- Tested with cri-o

### Changes
- update calico to v3.17.0
- update metrics-server to v0.4.1

### Fixed
- set cgroup according to host implementation

## 0.0.5
### Changes
- Tests with containerd on debian

## 0.0.4
### Added
- containerd compatibility

## 0.0.3
### Added
- Verify yum gpg key by default
- Ability to disable yum gpg keys in case a corporate proxy prevents getting them
- Debian distributions are now supported and tested on debian buster

### Changes
- Iptables rules aren't managed anymore by this role
- Lowered csr appoval for kubelet verbosity
- BREAKING : Flattened structure centos.kubernetes.repo to kubernetes_yum_repo variable
- BREAKING : Renamed kubernetes_pki vars for translation


## 0.0.2
### Added
- Ability to use a custom CA chain

### Fixed
- Cordon doesn't fail anymore when pods use local storage


## 0.0.1
### Added
- Supports kubernetes version 1.19.2
- Supports kubernetes version 1.19.3
- Calico cni deployment
- Flannel cni deployment
- metrics-server deployment
- kubelet tls bootstrap
- ipvs configuration for kube-proxy
- Client x509 authentification configuration
- Custom cluster AC configuration

### Changes
- BREAKING: rename k8s_kubeconfig var to kubernetes_kubeconfig_file

### Fixed


molecule/tests/conftest.py
"""PyTest Fixtures."""
from __future__ import absolute_import

import os

import pytest


def pytest_runtest_setup(item):
    """Run tests only when under molecule with testinfra installed."""
    try:
        import testinfra
    except ImportError:
        pytest.skip("Test requires testinfra", allow_module_level=True)

    if "MOLECULE_INVENTORY_FILE" in os.environ:
        pytest.testinfra_hosts = testinfra.utils.ansible_runner.AnsibleRunner(
            os.environ["MOLECULE_INVENTORY_FILE"]
        ).get_hosts("all")
    else:
        pytest.skip(
            "Test should run only from inside molecule.", allow_module_level=True
        )


molecule/tests/test_kubectl.py
"""Role testing files using testinfra."""
import re


def test_kubectl_package(host):
    kubectl = host.package("kubectl")
    assert kubectl.is_installed
    assert kubectl.version.startswith("1.19")

def test_kubectl_config(host):
    ansible_vars =  host.ansible.get_variables()

    if "kubernetes_masters" in ansible_vars['group_names']:
        config = host.file("/root/.kube/config")
        assert config.exists
        assert config.user == "root"
        assert config.group == "root"
        assert config.mode == 0o600
    else:
        config = host.file("/root/.kube/config")
        assert not config.exists

def test_kubectl_connection(host):
    ansible_vars =  host.ansible.get_variables()

    if "kubernetes_masters" in ansible_vars['group_names']:
        kubectl = host.run('kubectl version')
        assert kubectl.rc == 0
    else:
        kubectl =  host.run('kubectl version')
        assert kubectl.rc == 1

def test_kubectl_client_server_version(host):
    ansible_vars =  host.ansible.get_variables()

    if "kubernetes_masters" in ansible_vars['group_names']:
        version_raw = re.sub(r"[CS][a-z]+ Version: ", "", host.run('kubectl version --short').stdout)
        versions = version_raw.split('\n')

        assert versions[0] == versions[1]


molecule/commun/tests/conftest.py
"""PyTest Fixtures."""
from __future__ import absolute_import

import os

import pytest


def pytest_runtest_setup(item):
    """Run tests only when under molecule with testinfra installed."""
    try:
        import testinfra
    except ImportError:
        pytest.skip("Test requires testinfra", allow_module_level=True)

    if "MOLECULE_INVENTORY_FILE" in os.environ:
        pytest.testinfra_hosts = testinfra.utils.ansible_runner.AnsibleRunner(
            os.environ["MOLECULE_INVENTORY_FILE"]
        ).get_hosts("all")
    else:
        pytest.skip(
            "Test should run only from inside molecule.",
            allow_module_level=True
        )


molecule/commun/env_ugi.sh
DOCKER_REGISTRY_MIRROR=http://10.0.4.40:5000
QUAY_REGISTRY_MIRROR=10.0.4.40:5002
K8S_REGISTRY_MIRROR=10.0.4.40:5001
INSECURE_REGISTRIES=["10.0.4.40:5000","10.0.4.40:5001","10.0.4.40:5002"]
CENTOS_BASE_MIRROR=http://10.0.4.40:8081/repository/centos-base/centos
DEBIAN_BASE_MIRROR=http://10.0.4.40:8081/repository/debian-proxy
DOCKER_YUM_MIRROR=http://10.0.4.40:8081/repository/docker-yum
DOCKER_APT_MIRROR=http://10.0.4.40:8081/repository/docker-debian
K8S_YUM_MIRROR=http://10.0.4.40:8081/repository/kube-yum/
K8S_APT_MIRROR=http://10.0.4.40:8081/repository/kubernetes-deb
CRIO_BASE_URL=http://10.0.4.40:8081/repository/download-opensuse/
CONTAINERD_REGISTRY_MIRRORS='[{"name":"10.0.4.40:5000","endpoint":["http://10.0.4.40:5000"]},{"name":"10.0.4.40:5001","endpoint":["http://10.0.4.40:5001"]},{"name":"10.0.4.40:5002","endpoint":["http://10.0.4.40:5002"]},{"name":"k8s.gcr.io","endpoint":["http://10.0.4.40:5001"]},{"name":"quay.io","endpoint":["http://10.0.4.40:5002"]},{"name":"docker.io","endpoint":["http://10.0.4.40:5000"]}]'
CONTAINERD_REGISTRY_CONFIGS=[{"name":"10.0.4.40:5000","tls":{"insecure_skip_verify":true}}]
CRIO_REGISTRIES=[{"prefix":"10.0.4.40:5000","insecure":true,"location":"10.0.4.40:5000"},{"prefix":"10.0.4.40:5001","insecure":true,"location":"10.0.4.40:5001"},{"prefix":"10.0.4.40:5002","insecure":true,"location":"10.0.4.40:5002"},{"prefix":"quay.io","insecure":true,"blocked":false,"location":"10.0.4.40:5002"},{"prefix":"k8s.gcr.io","insecure":true,"location":"10.0.4.40:5001"},{"prefix":"docker.io","insecure":true,"location":"10.0.4.40:5000"}]


molecule/commun/env_ci_pe.sh
DOCKER_REGISTRY_MIRROR=https://docker-dev-virtual.repository.pole-emploi.intra
QUAY_REGISTRY_MIRROR=docker-quay-proxy.repository.pole-emploi.intra
K8S_REGISTRY_MIRROR=docker-k8s-gcr-proxy.repository.pole-emploi.intra
INSECURE_REGISTRIES=["docker-dev-virtual.repository.pole-emploi.intra", "docker-quay-proxy.repository.pole-emploi.intra", "docker-k8s-gcr-proxy.repository.pole-emploi.intra"]
CENTOS_BASE_MIRROR=https://repository.pole-emploi.intra/artifactory/rpm-base-centos-proxy
DEBIAN_BASE_MIRROR=http://artefact-repo.pole-emploi.intra/artifactory/deb.debian.org-Proxy
DOCKER_YUM_MIRROR=http://artefact-repo.pole-emploi.intra/artifactory/DebianDocker-proxy
DOCKER_APT_MIRROR=http://artefact-repo.pole-emploi.intra/artifactory/DebianDocker-proxy
K8S_YUM_MIRROR=https://repository.pole-emploi.intra/artifactory/rpm-kubernetes-proxy/
K8S_APT_MIRROR=http://artefact-repo.pole-emploi.intra/artifactory/DebianKubernetes-Proxy
CONTAINERD_REGISTRY_MIRRORS='[{"name":"docker-dev-virtual.repository.pole-emploi.intra","endpoint":["https://docker-dev-virtual.repository.pole-emploi.intra"],"skip_ssl_verify":true},{"name":"docker-quay-proxy.repository.pole-emploi.intra","endpoint":["https://docker-quay-proxy.repository.pole-emploi.intra"],"skip_ssl_verify":true},{"name":"docker-k8s-gcr-proxy.repository.pole-emploi.intra","endpoint":["https://docker-k8s-gcr-proxy.repository.pole-emploi.intra"],"skip_ssl_verify":true},{"name":"k8s.gcr.io","endpoint":["https://docker-k8s-gcr-proxy.repository.pole-emploi.intra"],"skip_ssl_verify":true},{"name":"docker.io","endpoint":["https://docker-dev-virtual.repository.pole-emploi.intra"]}]'


molecule/commun/env_pe_distant.sh
DOCKER_REGISTRY_MIRROR=https://docker-dev-virtual.repository.pole-emploi.intra
QUAY_REGISTRY_MIRROR=docker-quay-proxy.repository.pole-emploi.intra
K8S_REGISTRY_MIRROR=docker-k8s-gcr-proxy.repository.pole-emploi.intra
INSECURE_REGISTRIES=["docker-dev-virtual.repository.pole-emploi.intra", "docker-quay-proxy.repository.pole-emploi.intra", "docker-k8s-gcr-proxy.repository.pole-emploi.intra"]
CENTOS_BASE_MIRROR=https://repository.pole-emploi.intra/artifactory/rpm-base-centos-proxy
DEBIAN_BASE_MIRROR=http://artefact-repo.pole-emploi.intra/artifactory/deb.debian.org-Proxy
DOCKER_YUM_MIRROR=http://artefact-repo.pole-emploi.intra/artifactory/DebianDocker-proxy
DOCKER_APT_MIRROR=http://artefact-repo.pole-emploi.intra/artifactory/DebianDocker-proxy
K8S_YUM_MIRROR=https://repository.pole-emploi.intra/artifactory/rpm-kubernetes-proxy/
K8S_APT_MIRROR=http://artefact-repo.pole-emploi.intra/artifactory/DebianKubernetes-Proxy
CONTAINERD_REGISTRY_MIRRORS='[{"name":"docker-dev-virtual.repository.pole-emploi.intra","endpoint":["https://docker-dev-virtual.repository.pole-emploi.intra"]},{"name":"docker-quay-proxy.repository.pole-emploi.intra","endpoint":["https://docker-quay-proxy.repository.pole-emploi.intra"]},{"name":"docker-k8s-gcr-proxy.repository.pole-emploi.intra","endpoint":["https://docker-k8s-gcr-proxy.repository.pole-emploi.intra"]},{"name":"k8s.gcr.io","endpoint":["https://docker-k8s-gcr-proxy.repository.pole-emploi.intra"]},{"name":"docker.io","endpoint":["https://docker-dev-virtual.repository.pole-emploi.intra"]}]'
CONTAINERD_REGISTRY_CONFIGS=[{"name":"docker-dev-virtual.repository.pole-emploi.intra","tls":{"insecure_skip_verify":true}},{"name":"docker-quay-proxy.repository.pole-emploi.intra","tls":{"insecure_skip_verify":true}},{"name":"docker-k8s-gcr-proxy.repository.pole-emploi.intra","tls":{"insecure_skip_verify":true}},{"name":"k8s.gcr.io","tls":{"insecure_skip_verify":true}},{"name":"docker.io","tls":{"insecure_skip_verify":true}}]
REMOTE_HOST=true
PROVIDER_ARGS=['host="iugi3000-HP-ProDesk-400-G4-MT.dgasi.pole-emploi.intra"', 'username="libvirt-user"']
SSH_CONFIG=ssh_config_pe
SSH_ARGS=-o ProxyCommand="ssh -W %h:%p -q -l libvirt-user iugi3000-HP-ProDesk-400-G4-MT.dgasi.pole-emploi.intra"
http_proxy
https_proxy
HTTP_PROXY
HTTPS_PROXY
no_proxy
NO_PROXY


molecule/commun/env_ugi_distant.sh
DOCKER_REGISTRY_MIRROR=http://10.0.4.40:5000
QUAY_REGISTRY_MIRROR=10.0.4.40:5002
K8S_REGISTRY_MIRROR=10.0.4.40:5001
INSECURE_REGISTRIES=["10.0.4.40:5000","10.0.4.40:5001","10.0.4.40:5002"]
CENTOS_BASE_MIRROR=http://10.0.4.40:8081/repository/centos-base/centos
DEBIAN_BASE_MIRROR=http://10.0.4.40:8081/repository/debian-proxy
DOCKER_YUM_MIRROR=http://10.0.4.40:8081/repository/docker-yum
DOCKER_APT_MIRROR=http://10.0.4.40:8081/repository/docker-debian
K8S_YUM_MIRROR=http://10.0.4.40:8081/repository/kube-yum/
K8S_APT_MIRROR=http://10.0.4.40:8081/repository/kubernetes-deb
CRIO_BASE_URL=http://10.0.4.40:8081/repository/download-opensuse/
CONTAINERD_REGISTRY_MIRRORS='[{"name":"10.0.4.40:5000","endpoint":["http://10.0.4.40:5000"]},{"name":"10.0.4.40:5001","endpoint":["http://10.0.4.40:5001"]},{"name":"10.0.4.40:5002","endpoint":["http://10.0.4.40:5002"]},{"name":"docker.io","endpoint":["http://10.0.4.40:5000"]}]'
CONTAINERD_REGISTRY_CONFIGS=[{"name":"10.0.4.40:5000","tls":{"insecure_skip_verify":true}}]
CRIO_REGISTRIES=[{"prefix":"10.0.4.40:5000","insecure":true,"location":"10.0.4.40:5000"},{"prefix":"10.0.4.40:5001","insecure":true,"location":"10.0.4.40:5001"},{"prefix":"10.0.4.40:5002","insecure":true,"location":"10.0.4.40:5002"},{"prefix":"quay.io","insecure":true,"blocked":false,"location":"10.0.4.40:5002"},{"prefix":"k8s.gcr.io","insecure":true,"location":"10.0.4.40:5001"},{"prefix":"docker.io","insecure":true,"location":"10.0.4.40:5000"}]
REMOTE_HOST=true
PROVIDER_ARGS=['host="kvm"', 'username="libvirt-user"']
SSH_CONFIG=ssh_config_ugi
SSH_ARGS=-o ProxyCommand="ssh -W %h:%p -q -l libvirt-user kvm"
